docs/config_options.rst

```diff
@@ -3,43 +3,143 @@ Configuration Options
+     - Most log messages are sent here.
+     - ``logging.Logger``
+     -
+     -
+
+       .. _config_logprint:
+   * - ``logprint`` :ref:`🔗 <config_logprint>`
+     - Messages about available tilesources are sent here.
+     - ``logging.Logger``
+     -
+     -
+
+       .. _config_cache_backend:
+   * - ``cache_backend`` :ref:`🔗 <config_cache_backend>`
+     - String specifying how tiles are cached.  If memcached is not available for any reason, the python cache is used instead.
+     - ``str: "python" | "memcached"``  # What about redis?
+     - ``False``
+     - ``"python"``
+
+       .. _config_cache_python_memory_portion:
+   * - ``cache_python_memory_portion`` :ref:`🔗 <config_cache_python_memory_portion>`
+     - If tiles are cached with python, the cache is sized so that it is expected to use less than 1 / (``cache_python_memory_portion``) of the available memory.
+     - ``int``
+     -
+     -
+
+       .. _config_cache_memcached_url:
+   * - ``cache_memcached_url`` :ref:`🔗 <config_cache_memcached_url>`
+     - If tiles are cached in memcached, the url or list of urls where the memcached server is located.
+     - ``str | List[str]``
+     - ``False``
+     - ``"127.0.0.1"``
+
+       .. _config_cache_memcached_username:
+   * - ``cache_memcached_username`` :ref:`🔗 <config_cache_memcached_username>`
+     - A username for the memcached server.
+     -
+     -
+     - ``None``
+
+       .. _config_cache_memcached_password:
+   * - ``cache_memcached_password`` :ref:`🔗 <config_cache_memcached_password>`
+     - A password for the memcached server.
+     -
+     -
+     - ``None``
+
+       .. _config_cache_redis_url:
+   * - ``cache_redis_url`` :ref:`🔗 <config_cache_redis_url>`
+     - If tiles are cached in redis, the url or list of urls where the redis server is located.
+     - ``str | List[str]``
+     -
+     - ``"127.0.0.1:6379"``
+
+       .. _config_cache_redis_username:
+   * - ``cache_redis_username`` :ref:`🔗 <config_cache_redis_username>`
+     - A username for the redis server.
+     -
+     -
+     - ``None``
+
+       .. _config_cache_redis_password:
+   * - ``cache_redis_password`` :ref:`🔗 <config_cache_redis_password>`
+     - A password for the redis server.
+     -
+     -
+     - ``None``
+
+       .. _config_cache_tilesource_memory_portion:
+   * - ``cache_tilesource_memory_portion`` :ref:`🔗 <config_cache_tilesource_memory_portion>`
+     - Tilesources are cached on open so that subsequent accesses can be faster.  These use file handles and memory.  This limits the maximum based on a memory estimation and using no more than 1 / (``cache_tilesource_memory_portion``) of the available memory.
+     - ``int``
+     -
+     -
+
+       .. _config_cache_tilesource_maximum:
+   * - ``cache_tilesource_maximum`` :ref:`🔗 <config_cache_tilesource_maximum>`
+     - If this is non-zero, this further limits the number of tilesources than can be cached to this value.
+     -
+     -
+     -
```

Comment:
The minimum is always 3 (unless cache_sources is False), regardless of either this or the memory portion setting.

Reply:
Could you clarify this one? What is the default value for `cache_tilesource_maximum`?

Reply:
The default is `0` which is the signifier that `cache_tilesource_memory_portion` determines the number of sources that will be cached.  If >0 the cache will be the smaller of the value computed for the memory portion and this value (but always at least 3).

Reply:
If the user specifies `cache_tilesource_maximum=2`, is the behavior the same as specifying 1 or 0? Does this value only work if greater than 3?

Reply:
Specifying 1 or 2 is the same as specifying 3.

Reply:
Thanks. Added in [4465d0c](https://github.com/girder/large_image/pull/1597/commits/4465d0c81b67d58e243766b6441257327a2cfcdf)

Reply:
The primary reason to set this is to limit memory use.  In the (expectedly) rare cases where you don't want caching, you'd set `cache_sources` to False.

Reply:
In some sense the minimum of 3 should be considered an implementation detail.  I could imagine a refactor where a value of 1 or 2 could be honored.

Internally, we treat styled sources as a separate cache entry to unstyled sources, which means that, currently for the cache to actually work for styled sources we need to keep both the styled and unstyled source in the cache (hence a minimum of 2).  Further, when opening a multi-source source that uses styled base sources, a value of 2 can cause the cache to cycle and act like it is functionally not there (and thus we get the minimum of 3).  In practice, you could have a multi-source source that could be itself styled and could have many styled base sources, then you'd really want this number to be 1 + (1 if multi-source source is styled else 0) + number of distinct based sources + number of distinct style/base source combinations, , but this isn't always practicable (e.g., we've seen multi-source sources with 50,000 base sources).

We pick 3 as a minimum with a reasonable compromise for honoring the intent and not breaking caching in common situations.